firmware/chipignite/util/caravel/hk.py

```diff
@@ -0,0 +1,155 @@
+from .find_ftdi import find_ftdi
+from .gpio import Gpio
+from .defs import *
+from pyftdi.spi import SpiController
+import time
+
+
+class HKSpi:
+    def __init__(self, ftdi_device: str=None):
+        if ftdi_device == None:
+            ftdi_device = find_ftdi()
+        
+        self.spi = SpiController(cs_count=1)
+        self.spi.configure(ftdi_device)
+
+        self.raw_gpio = self.spi.get_gpio()
+        self.gpio = Gpio(self.raw_gpio)
+        self.led = self.gpio.get_pin(10, value=0)
+        self.led2 = self.gpio.get_pin(11, value=0)
+        # self.mr = self.gpio.get_pin(9)
+
+        # Important, disable UART by driving UART_EN high _before_ opening SPI
+        self.uart_en = self.gpio.get_pin(8)
+
+        self.slave = self.spi.get_port(cs=0, freq=1E6, mode=0)
+
+    def identify(self):
+        print("Caravel data:")
+
+        mfg = self.slave.exchange([CARAVEL_STREAM_READ, 0x01], 2)
+        print("   mfg        = {:04x}".format(int.from_bytes(mfg, byteorder='big')))
+        
+        product = self.slave.exchange([CARAVEL_REG_READ, 0x03], 1)
+        print("   product    = {:02x}".format(int.from_bytes(product, byteorder='big')))
+        
+        data = self.slave.exchange([CARAVEL_STREAM_READ, 0x04], 4)
+        print("   project ID = {:08x}".format(int('{:032b}'.format(int.from_bytes(data, byteorder='big'))[::-1], 2)))
+        
+        if int.from_bytes(mfg, byteorder='big') != 0x0456:
+            print("Incorrect MFG value, expected 0x0456.")
+            exit(2)
+
+    def get_status(self):
+        return int.from_bytes(self.slave.exchange([CARAVEL_PASSTHRU, CMD_READ_STATUS],1), byteorder='big')
+
+    def print_status(self):
+        print("status = 0x{:02x}".format(self.get_status()))
+
+    def print_long_status(self):
+        jedec = self.flash_read_jedec()
+        if jedec[0] == int('bf', 16):
+            print("changing cmd values...")
+            print("status reg_1 = {}".format(hex(self.get_status())))
+        else:
+            print("status reg_1 = {}".format(hex(self.get_status())))
+            status = self.slave.exchange([CARAVEL_PASSTHRU, 0x35], 1)
+            print("status reg_2 = {}".format(hex(int.from_bytes(status, byteorder='big'))))
+            # print("status = {}".format(hex(from_bytes(slave.exchange([CMD_READ_STATUS], 2)[1], byteorder='big'))))
+
+
+
+    def is_busy(self):
+        return self.get_status() & SR_WIP
+
+    def read_reg(self, reg: int):
+        return self.slave.exchange([CARAVEL_REG_READ, reg], 1)
+    
+    def write_reg(self, reg: int, value: int):
+        self.slave.exchange([CARAVEL_REG_WRITE, reg, value], 1)
```

Comment:
Is the `,1` at the end of this intentional for a *write*?

Reply:
Great question, I'm not actually sure. Half of this was copy-pasted from the old version, and some was debugging a few issues with how it communicated with a slightly different flash chip. That looks like it might have been copy-pasted from line 66, but I'm not entirely sure